### DISCUSSION TITLE
Narrow doesn't work on class properties

### DISCUSSION THREAD

**Discussion by DetachHead** [millsp/ts-toolbelt#212] [Q&A] [answered]
#### Describe the bug

`Narrow` doesn't work on class properties
#### Reproduce the bug
```ts
interface Foo {
    bar: Narrow<{hi: 'there'}>
}

class Bar implements Foo {
    bar = {hi: 'there'} //error: Type 'string' is not assignable to type 'NarrowRaw<"there">'.
}
```

#### Expected behavior

the value of `bar` is narrowed to type `{hi: 'there'}`
#### Screenshots

![image](https://user-images.githubusercontent.com/57028336/109960269-530c0480-7d34-11eb-8b22-34c8192d8425.png)
#### Additional context

ts-toolbelt 9.5.3
typescript 4.2.0

**Comment by millsp** [comment, marked as answer]
`Narrow` only works on generic types, that is to say on parameter input:

```ts
class C<B> {
    b: F.Narrow<B>;

    constructor(b: F.Narrow<B>) {
        this.b = b
    }
}

const c = new C({a: 1})
```

or write a helper

```ts
const narrow = <T>(thing: Narrow<T>) => thing

interface Foo {
    bar: {hi: 'there'}
}

class Bar implements Foo {
    bar = narrow({hi: 'there'})
}
```


Hope this will help

**Reply by DetachHead** [reply]
is this a typescript limitation or an intentional decision?

i'm having trouble understanding how this type works tbh, how exactly does it force types at the call site to be narrowed?

**Reply by millsp** [reply]
It is some kind of limitation, we cannot narrow other things than a generic type. It has to pass through a function first. So you could as well write a very useful little utility like:

```ts
const narrow = <T>(thing: Narrow<T>) => thing

interface Foo {
    bar: {hi: 'there'}
}

class Bar implements Foo {
    bar = narrow({hi: 'there'})
}
```

**Reply by millsp** [reply]
Or you can always use `as const` assertions, but this has the inconvenient to make things deeply readonly.
